fix(upgrade): prevent longhorn settings be reset to the value stored in Upgrade CR unexpectedly

wheatdog opened the pull request:
<!-- 
!IMPORTANT!
Please do not create a Pull Request without creating an issue first.
-->

#### Problem:
<!-- Explain the problem you aim to resolve in this PR. -->
When restarting the harvester controller, it will pick up the latest completed Upgrade CR and `Replica Replenishment Wait Interval` from longhorn settings will be reset to the value stored in the annotation of Upgrade CR. But the value stored in the annotation might already be stale.

#### Solution:
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->
To prevent unexpected restoring again, a new annotation `harvesterhci.io/longhorn-settings-restored` is introduced to be set right after the corresponding settings are restored. If this annotation is set, the longhorn settings value restoration would be skipped.

#### Related Issue(s):
<!--
Use `Issue #<issue number>` or `Issue harvester/harvester#<issue number>` or `Issue (paste link of issue)`. DON'T use `Fixes #<issue number>` or `Fixes (paste link of issue)`, as it will automatically close the linked issue when the PR is merged.
-->
#8934 

#### Test plan:
<!-- Describe the test plan by steps. -->

1. Provision a harvester cluster (at least 2 nodes)
2. Perform an upgrade (2 scenarios)
    1. from version **without this fix** to version **with this fix**
    2. from version **with this fix** to version **with this fix**
3. Modify Replica Replenishment Wait Interval 600 -> 60.
4. Restart harvester controller
    ```
    kubectl rollout -n harvester-system restart deployment/harvester
    ```
5. Replica Replenishment Wait Interval should still be 60
6. The new annotation should be set from Upgrade CR


#### Additional documentation or context

n/a

wheatdog committed:
fix(upgrade): introduce cleanupPlanCompletedAnnotation for idempotency

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
Revert "fix(upgrade): introduce cleanupPlanCompletedAnnotation for idempotency"

This reverts commit 43a8994b1fc2a1e66a5c1534c4b4366a21698978.

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
fix(upgrade): delete annotations after restoring

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
test(upgrade): delete annotations after restoring

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
Rename the new test case

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
chores: log with info log level

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
feat(upgrade): introduce longhornSettingsRestoredAnnotation instead of removing existings

Signed-off-by: Tim Liou <tim.liou@suse.com>

FrankYang0529 reviewed (commented):
Over all LGTM. Leave some minor comments.

w13915984028 reviewed (commented):
A few additional questions, thanks.

wheatdog committed:
chores: remove redundant code

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
fix: skip IsNotFound error while deleting

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
fix: check deletionTimestamp before deleting

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
fix: directly book the longhornSettingsRestoredAnnotation update

To make this annotation land ASAP, and leave the rest of lines be
handled by next OnChange call

Signed-off-by: Tim Liou <tim.liou@suse.com>

w13915984028 reviewed (approved):
LGTM, thanks for the enhancement.
